File: config-auh-core-basic/assets/auth-core-basic-template/src/user/usecase/find-by-email.usecase.ts
import { Result, UseCase } from "__SHARED_PACKAGE_NAME__";
import { UserDTO } from "../dto";
import { FindUserByEmailQuery } from "../provider";

export interface FindUserByEmailOut extends UserDTO {}

export class FindUserByEmailUseCase
	implements UseCase<string, FindUserByEmailOut>
{
	constructor(private readonly findByEmail: FindUserByEmailQuery) {}

	async execute(email: string): Promise<Result<FindUserByEmailOut>> {
		const userResult = await this.findByEmail.execute(email);
		if (userResult.isFailure) {
			return userResult.withFail;
		}

		return Result.ok(userResult.instance);
	}
}

File: config-auth-core-basic-kt/assets/auth-core-basic-template-kt/src/main/kotlin/com/example/auth/user/usecase/FindByEmailUseCase.kt
package com.example.auth.user.usecase

import com.example.auth.user.dto.UserDto
import com.example.auth.user.query.FindUserByEmailQuery

class FindByEmailUseCase(
    private val findUserByEmail: FindUserByEmailQuery
) {
    suspend fun execute(email: String): Result<UserDto> {
        val user = findUserByEmail.findByEmail(email)
            ?: return Result.failure(IllegalArgumentException("USER_NOT_FOUND"))
        return Result.success(user)
    }
}
